base/promotion.jl

```diff
@@ -102,7 +102,7 @@ function typejoin(@nospecialize(a), @nospecialize(b))
             vars = []
             for i = 1:n
                 ai, bi = a.parameters[i], b.parameters[i]
-                if ai === bi || (isa(ai,Type) && isa(bi,Type) && typeseq(ai,bi))
+                if ai === bi || (isa(ai,Type) && isa(bi,Type) && ai == bi)
```

Comment:
Doesn't this not cause specialization of `==` on all argument types?

Reply:
I'll change this call site just in case it helps, but we're inevitably going to end up calling `==` and `isequal` on types anyway. We could even consider removing the `==` method for types, but `==` would still be called and hit the fallback definition.

Reply:
Calling `==` on types is fine, but this function is deliberately annotated `@nospecialize`, so we should try to avoid having it call back into the compiler internally to avoid defeating the purpose.

Reply:
We don't specialize functions on `Type` anyways, but we could mark `==` no-specialize too